### PR TITLE
[JENKINS-39602] upgrade pac4j to 1.9.9

### PR DESCRIPTION
**Pull request by kuisathaverat**
[JENKINS-39602](https://issues.jenkins-ci.org/browse/JENKINS-39602)

* Upgrade the pac4j library to 1.9.9
* Change the minimum Java version supported to Java 8 (required by pac4j)
* Change the minimum Jenkins core to 2.60 (first full Java 8 LTS )
* Increase the major version to 1.0.0 
* Update Parent pom (2.32)
* Normalize the code format on all files.

@reviewbybees 

I started from this [PR](https://github.com/jenkinsci/saml-plugin/pull/24) but finally, I upgraded to the last 1.XX version and this implies tons of changes.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -25,11 +25,11 @@ under the License.
   <parent>
     <groupId>org.jenkins-ci.plugins</groupId>
     <artifactId>plugin</artifactId>
-    <version>2.17</version>
+    <version>2.32</version>
   </parent>
 
   <artifactId>saml</artifactId>
-  <version>0.15-SNAPSHOT</version>
+  <version>1.00-SNAPSHOT</version>
```
Comment (recena) [inline pom.xml]: @kuisathaverat Do you mean `1.0.0-SNAPSHOT` or `1.0-SNAPSHOT`? `1.00` sounds like a weird version naming.

File: pom.xml [hunk 2]
```diff
@@ -79,21 +80,44 @@ under the License.
...
     <dependency>
       <groupId>org.jenkins-ci.plugins</groupId>
       <artifactId>mailer</artifactId>
       <version>1.18</version>
       <scope>compile</scope>
     </dependency>
+    <dependency>
+      <groupId>net.shibboleth.utilities</groupId>
+      <artifactId>java-support</artifactId>
+      <version>7.2.0</version>
```
Comment (amuniz) [inline pom.xml]: What is this dependency for?

File: src/main/java/org/jenkinsci/plugins/saml/BundleKeyStore.java [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+import java.util.logging.Logger;
+
+/**
+ * Pac4j requires to set a keystore for encryption operations,
+ * the plugin generate an automatic keystore or it it is not possible uses a keystore bundle on the plugin.
+ * The generated key is valid for a day, when expires it is generated a new one on the same keystore.
+ * A new key store is created when you restart Jenkins or if is not possible to access to the created.
+ * @see <a href="http://www.pac4j.org/1.9.x/docs/clients/saml.html">pac4j - Authentication mechanisms: SAML</a>
+ */
+public class BundleKeyStore {
```
Comment (amuniz) [inline src/main/java/org/jenkinsci/plugins/saml/BundleKeyStore.java]: How is this related to JENKINS-39602? Does pac4j:1.9.9 require it?

File: src/main/java/org/jenkinsci/plugins/saml/BundleKeyStore.java [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+
+    /**
+     * initialized the keystore, it tries to create a keystore in a file,
+     * if it fails load the settings of the demo keystore.
+     */
+    public synchronized void init() {
+        try {
+
+            if(keystore == null){
+                keystore = File.createTempFile("saml-jenkins-keystore-", ".jks");
```
Comment (amuniz) [inline src/main/java/org/jenkinsci/plugins/saml/BundleKeyStore.java]: Is it safe to put the keystore in `/tmp`?

File: src/main/java/org/jenkinsci/plugins/saml/BundleKeyStore.java [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+     * @return the keystore loaded.
+     * @throws KeyStoreException        @see KeyStoreException
+     * @throws IOException              @see IOException
+     * @throws CertificateException     @see CertificateException
+     * @throws NoSuchAlgorithmException @see NoSuchAlgorithmException
+     */
+    private KeyStore loadKeyStore(File keystore, String password)
+            throws KeyStoreException, IOException, CertificateException, NoSuchAlgorithmException {
+        KeyStore ks = KeyStore.getInstance(KeyStore.getDefaultType());
+        try{
```
Comment (amuniz) [inline src/main/java/org/jenkinsci/plugins/saml/BundleKeyStore.java]: You can use try-with-resources here for clarity.

```java
try (InputStream in = new FileInputStream(keystore)) {
      ks.load(in, password.toCharArray());
} catch (Exception e) { // better to catch something more specific
     ks = initKeyStore(keystore, password);
}
```

File: src/main/java/org/jenkinsci/plugins/saml/OpenSAMLWrapper.java [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+
+import java.io.File;
+import java.io.FileOutputStream;
+import java.security.KeyPairGenerator;
+import java.security.KeyStore;
+import java.security.PrivateKey;
+import java.security.SecureRandom;
+import java.security.cert.X509Certificate;
+import java.util.Date;
+*/
```
Comment (amuniz) [inline src/main/java/org/jenkinsci/plugins/saml/OpenSAMLWrapper.java]: Remove this.

File: src/main/java/org/jenkinsci/plugins/saml/OpenSAMLWrapper.java [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+            thread.setContextClassLoader(InitializationService.class.getClassLoader());
+            try {
+                InitializationService.initialize();
+                ret = process();
+            } finally {
+                LOG.finest("reset TCCL");
+                thread.setContextClassLoader(loader);
+            }
+        } catch (InitializationException e) {
+            //FIXME [kuisathaverat] throw exception.
```
Comment (amuniz) [inline src/main/java/org/jenkinsci/plugins/saml/OpenSAMLWrapper.java]: Exception is already thrown, is this FIXME still valid?

File: src/main/java/org/jenkinsci/plugins/saml/OpenSAMLWrapper.java [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+        return new J2EContext(request, response);
+    }
+
+
+    /**
+     * @return a SAML2Client object to interact with the IdP service.
+     */
+    protected SAML2Client createSAML2Client() {
+        //FIXME [kuisathaverat] Throws NPE I do not like it
+        Preconditions.checkNotNull(samlPluginConfig.getIdpMetadata());
```
Comment (amuniz) [inline src/main/java/org/jenkinsci/plugins/saml/OpenSAMLWrapper.java]: If the client is not usable without the IdP metadata, then just make it required (and add a validation check) in the configuration and assume it is non-null anywhere else.

File: src/main/java/org/jenkinsci/plugins/saml/OpenSAMLWrapper.java [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+            // reference) is set, include it in the request to the IdP, and request
+            // that the IdP uses exact matching for authentication types
+            if (samlPluginConfig.getAuthnContextClassRef() != null) {
+                config.setAuthnContextClassRef(samlPluginConfig.getAuthnContextClassRef());
+                config.setComparisonType("exact");
+            }
+        }
+
+        final SAML2Client saml2Client = new SAML2Client(config);
+        saml2Client.setCallbackUrl(samlPluginConfig.getConsumerServiceUrl());
```
Comment (amuniz) [inline src/main/java/org/jenkinsci/plugins/saml/OpenSAMLWrapper.java]: What's this configuration for? AFAIK the callback URL is set on the IdP (linked to a service provider ID).

File: src/main/java/org/jenkinsci/plugins/saml/OpenSAMLWrapper.java [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+        saml2Client.init(createWebContext());
+        String spMetadata = saml2Client.getServiceProviderMetadataResolver().getMetadata();
```
Comment (amuniz) [inline src/main/java/org/jenkinsci/plugins/saml/OpenSAMLWrapper.java]: Is this supported by all SAML 2.0 IdPs? Does it mean there is no need to configure it in the UI anymore?

File: src/main/java/org/jenkinsci/plugins/saml/RewriteableStringResource.java [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.File;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.io.UnsupportedEncodingException;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+class RewriteableStringResource implements WritableResource {
```
Comment (amuniz) [inline src/main/java/org/jenkinsci/plugins/saml/RewriteableStringResource.java]: This looks like a general purpose class but in this plugin it is used for SAML metadata processing. Why not calling it appropriately? It is a bit confusing to me at least.

File: src/main/java/org/jenkinsci/plugins/saml/RewriteableStringResource.java [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+            throw new IOException("no data");
+        }
+        return new ByteArrayInputStream(string);
+    }
+
+    @Override
+    public File getFile() {
+        File temp = null;
+        try {
+            temp = File.createTempFile("jenkins-saml-", ".bin");
```
Comment (amuniz) [inline src/main/java/org/jenkinsci/plugins/saml/RewriteableStringResource.java]: How many times is this method called in a normal use case flow? Generating a temp file on every call seems a bit weird.

File: src/main/java/org/jenkinsci/plugins/saml/SamlAuthenticationToken.java [hunk 13]
```diff
@@ -34,37 +34,37 @@
...
+
+        return userDetails;
+    }
 
-  public String getCredentials() {
-    return "SAML does not use passwords";
-  }
+    public String getCredentials() {
+        return "SAML does not use passwords";
+    }
```
Comment (amuniz) [inline src/main/java/org/jenkinsci/plugins/saml/SamlAuthenticationToken.java]: Is this chunk all code style fixes? I don't see changes but I might be missing something.

File: src/main/java/org/jenkinsci/plugins/saml/SamlProfileWrapper.java [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+import java.util.logging.Logger;
+
+/**
+ * Process to response from the IdP to obtain the SAML2Profile of the user.
+ */
+public class SamlProfileWrapper extends OpenSAMLWrapper<SAML2Profile> {
+    private static final Logger LOG = Logger.getLogger(SamlProfileWrapper.class.getName());
+
+
+    public SamlProfileWrapper(SamlPluginConfig samlPluginConfig, StaplerRequest request, StaplerResponse response) {
```
Comment (amuniz) [inline src/main/java/org/jenkinsci/plugins/saml/SamlProfileWrapper.java]: Where is this class being instantiated?

File: src/main/java/org/jenkinsci/plugins/saml/BundleKeyStore.java [hunk 15]
```diff
@@ -0,0 +1,277 @@
...
+
+import sun.security.x509.AlgorithmId;
+import sun.security.x509.CertificateAlgorithmId;
+import sun.security.x509.CertificateSerialNumber;
+import sun.security.x509.CertificateValidity;
+import sun.security.x509.CertificateVersion;
+import sun.security.x509.CertificateX509Key;
+import sun.security.x509.X500Name;
+import sun.security.x509.X509CertImpl;
+import sun.security.x509.X509CertInfo;
```
Comment (amuniz) [inline src/main/java/org/jenkinsci/plugins/saml/BundleKeyStore.java]: All this classes are internal proprietary APIs and may be removed in a future release (as noted during compilation). Did you check if [bouncycastle-api-plugin](https://github.com/jenkinsci/bouncycastle-api-plugin/blob/06785880330f3c52113e5f18fbc13e9d46162de6/src/main/java/jenkins/bouncycastle/api/PEMEncodable.java) provides what you need here?

This is a :bug: IMO

**Comment by amuniz** [comment]
Checked that the ATH passes using this local SNAPSHOT.

File: pom.xml (hunk 1 above)
Comment (kuisathaverat) [inline pom.xml]: yep, it is wrong should be 1.0.0

File: src/main/java/org/jenkinsci/plugins/saml/BundleKeyStore.java (hunk 3 above)
Comment (kuisathaverat) [inline src/main/java/org/jenkinsci/plugins/saml/BundleKeyStore.java]: require a keystore, there is one bundled but is a demo keystore this class add a automatic keystore an key, it is more secure than the bundled one.

File: src/main/java/org/jenkinsci/plugins/saml/BundleKeyStore.java (hunk 4 above)
Comment (kuisathaverat) [inline src/main/java/org/jenkinsci/plugins/saml/BundleKeyStore.java]: it is a temp file, it the OS it is proper configured /tmp has the flag "t" , so only the user that creates it should read it, additionally, the password for the key store is a 256bytes password

File: src/main/java/org/jenkinsci/plugins/saml/OpenSAMLWrapper.java (hunk 7 above)
Comment (kuisathaverat) [inline src/main/java/org/jenkinsci/plugins/saml/OpenSAMLWrapper.java]: no, I have to remove it too

File: src/main/java/org/jenkinsci/plugins/saml/OpenSAMLWrapper.java (hunk 10 above)
Comment (kuisathaverat) [inline src/main/java/org/jenkinsci/plugins/saml/OpenSAMLWrapper.java]: It is the SP metadata not the IdP metadata, I think you understood Idp metadata

File: src/main/java/org/jenkinsci/plugins/saml/OpenSAMLWrapper.java (hunk 9 above)
Comment (kuisathaverat) [inline src/main/java/org/jenkinsci/plugins/saml/OpenSAMLWrapper.java]: Is the callback URL it is required, even do is on the IdP configured

File: src/main/java/org/jenkinsci/plugins/saml/RewriteableStringResource.java (hunk 11 above)
Comment (kuisathaverat) [inline src/main/java/org/jenkinsci/plugins/saml/RewriteableStringResource.java]: WDYM?

File: src/main/java/org/jenkinsci/plugins/saml/RewriteableStringResource.java (hunk 12 above)
Comment (kuisathaverat) [inline src/main/java/org/jenkinsci/plugins/saml/RewriteableStringResource.java]: It a new method required by pac4j, I do not see any call to it but it should be implemented.

File: src/main/java/org/jenkinsci/plugins/saml/SamlAuthenticationToken.java (hunk 13 above)
Comment (kuisathaverat) [inline src/main/java/org/jenkinsci/plugins/saml/SamlAuthenticationToken.java]: yep, I made a reformat there is almost three different styles of code with a different number of indent spaces.

File: src/main/java/org/jenkinsci/plugins/saml/SamlProfileWrapper.java (hunk 14 above)
Comment (kuisathaverat) [inline src/main/java/org/jenkinsci/plugins/saml/SamlProfileWrapper.java]: SamlSecurityRealm.doFinishLogin

File: src/main/java/org/jenkinsci/plugins/saml/BundleKeyStore.java (hunk 15 above)
Comment (kuisathaverat) [inline src/main/java/org/jenkinsci/plugins/saml/BundleKeyStore.java]: bouncycastle could work the method toCertificate return the Certificate that I need but I never used this library and I do not know how maintained it is, this certificates classes are used by keytool and it is difficult to think that could be removed, I am gonna try to use the bouncycastle library.

File: src/main/java/org/jenkinsci/plugins/saml/OpenSAMLWrapper.java (hunk 9 above)
Comment (amuniz) [inline src/main/java/org/jenkinsci/plugins/saml/OpenSAMLWrapper.java]: So, what happens if they don't match? Which is the one taking precedence (if any)?

File: src/main/java/org/jenkinsci/plugins/saml/OpenSAMLWrapper.java (hunk 10 above)
Comment (amuniz) [inline src/main/java/org/jenkinsci/plugins/saml/OpenSAMLWrapper.java]: Ahh, yeah. OK.

File: src/main/java/org/jenkinsci/plugins/saml/RewriteableStringResource.java (hunk 12 above)
Comment (amuniz) [inline src/main/java/org/jenkinsci/plugins/saml/RewriteableStringResource.java]: > I do not see any call to it 

How did you check? We should measure it somehow.

File: src/main/java/org/jenkinsci/plugins/saml/SamlProfileWrapper.java (hunk 14 above)
Comment (amuniz) [inline src/main/java/org/jenkinsci/plugins/saml/SamlProfileWrapper.java]: Mmm, I missed that diff because it was collapsed... checking now.

File: pom.xml (hunk 2 above)
Comment (kuisathaverat) [inline pom.xml]: pac4j and opemSAML but has a different version, this force to use 7.2

File: src/main/java/org/jenkinsci/plugins/saml/BundleKeyStore.java (hunk 15 above)
Comment (amuniz) [inline src/main/java/org/jenkinsci/plugins/saml/BundleKeyStore.java]: > I do not know how maintained it is

It is well maintained AFAICT, and used in other plugins.

Comment (kuisathaverat) [inline src/main/java/org/jenkinsci/plugins/saml/BundleKeyStore.java]: After taking a look to bouncycastle-api-plugin, it does not implement how to create a chain of certificates from a keypair nor create a certificate from a keypair I will have to implement it and the only way to create a certificate object is from an ASN1 input stream so I will have also to implement a weird code with temp files to create the certificates and the ASN1 input, so I think in this case bouncycastle-api-plugin does not add worth and make things more complicated,  bouncycastle-api-plugin is designed to import/export key and certificates from files or to files does not have anything to manage certificates with code (easy way) or keystores.

File: src/main/java/org/jenkinsci/plugins/saml/OpenSAMLWrapper.java (hunk 8 above)
Comment (kuisathaverat) [inline src/main/java/org/jenkinsci/plugins/saml/OpenSAMLWrapper.java]: pending of this https://issues.jenkins-ci.org/browse/JENKINS-40144

File: src/main/java/org/jenkinsci/plugins/saml/BundleKeyStore.java (hunk 5 above)
Comment (kuisathaverat) [inline src/main/java/org/jenkinsci/plugins/saml/BundleKeyStore.java]: ok

File: src/main/java/org/jenkinsci/plugins/saml/OpenSAMLWrapper.java (hunk 6 above)
Comment (kuisathaverat) [inline src/main/java/org/jenkinsci/plugins/saml/OpenSAMLWrapper.java]: ok, I create a Jira to do not forget it

File: src/main/java/org/jenkinsci/plugins/saml/BundleKeyStore.java (hunk 15 above)
Comment (amuniz) [inline src/main/java/org/jenkinsci/plugins/saml/BundleKeyStore.java]: > I will have to implement it

Yes. But I believe it is the right thing to do.

> the only way to create a certificate object is from an ASN1 input stream so I will have also to implement a weird code with temp files

[It does not seem like that](https://stackoverflow.com/questions/29852290/self-signed-x509-certificate-with-bouncy-castle-in-java).

In any case, using `sun.*` is not an option IMO ([more here](http://www.oracle.com/technetwork/java/faq-sun-packages-142232.html)).

File: src/main/java/org/jenkinsci/plugins/saml/BundleKeyStore.java [hunk 16]
```diff
@@ -0,0 +1,272 @@
...
+     * @throws IOException              @see IOException
+     * @throws CertificateException     @see CertificateException
+     * @throws NoSuchAlgorithmException @see NoSuchAlgorithmException
+     */
+    private KeyStore loadKeyStore(File keystore, String password)
+            throws KeyStoreException, IOException, CertificateException, NoSuchAlgorithmException {
+        KeyStore ks = KeyStore.getInstance(KeyStore.getDefaultType());
+        try (InputStream in = new FileInputStream(keystore)) {
+            ks.load(in, password.toCharArray());
+        } catch (Exception e) { // better to catch something more specific
```
Comment (amuniz) [inline src/main/java/org/jenkinsci/plugins/saml/BundleKeyStore.java]: `better to catch something more specific`

Well, I left the comment there for you to fix it 😄 
Don't catch `Exception`, catch whatever is thrown by the `try` block.

File: src/main/java/org/jenkinsci/plugins/saml/OpenSAMLWrapper.java (hunk 6 above)
Comment (amuniz) [inline src/main/java/org/jenkinsci/plugins/saml/OpenSAMLWrapper.java]: A JIRA to remove this comment? Really?

File: src/main/java/org/jenkinsci/plugins/saml/BundleKeyStore.java (hunk 15 above)
Comment (kuisathaverat) [inline src/main/java/org/jenkinsci/plugins/saml/BundleKeyStore.java]: with that samples I finally I could accomplish the task to move to the bouncy castle library

File: src/main/java/org/jenkinsci/plugins/saml/BundleKeyStore.java (hunk 16 above)
Comment (kuisathaverat) [inline src/main/java/org/jenkinsci/plugins/saml/BundleKeyStore.java]: done

File: src/main/java/org/jenkinsci/plugins/saml/RewriteableStringResource.java (hunk 12 above)
Comment (kuisathaverat) [inline src/main/java/org/jenkinsci/plugins/saml/RewriteableStringResource.java]: I found something about this method and some warnings on the console, it was been not used, but now I gonna use it to save the SP metadata on the Jenkins Home folder.

File: src/main/java/org/jenkinsci/plugins/saml/RewriteableStringResource.java (hunk 11 above)
Comment (kuisathaverat) [inline src/main/java/org/jenkinsci/plugins/saml/RewriteableStringResource.java]: I changed the name to something more accurate  SamlStringResource

File: src/main/java/org/jenkinsci/plugins/saml/OpenSAMLWrapper.java (hunk 10 above)
Comment (kuisathaverat) [inline src/main/java/org/jenkinsci/plugins/saml/OpenSAMLWrapper.java]: Here I will save the SP metadata to a file.

File: src/main/java/org/jenkinsci/plugins/saml/OpenSAMLWrapper.java (hunk 9 above)
Comment (kuisathaverat) [inline src/main/java/org/jenkinsci/plugins/saml/OpenSAMLWrapper.java]: I have to test it but I think that it depends on the IdP implementation, also I gonna take a look at the RFC of SAML to see what it is the recommendation.

File: src/main/java/org/jenkinsci/plugins/saml/OpenSAMLWrapper.java (hunk 7 above)
Comment (kuisathaverat) [inline src/main/java/org/jenkinsci/plugins/saml/OpenSAMLWrapper.java]: done

File: src/main/java/org/jenkinsci/plugins/saml/OpenSAMLWrapper.java (hunk 6 above)
Comment (kuisathaverat) [inline src/main/java/org/jenkinsci/plugins/saml/OpenSAMLWrapper.java]: no, only to remember that it is possible to add support to configure signature algorithms

File: src/main/java/org/jenkinsci/plugins/saml/BundleKeyStore.java (hunk 4 above)
Comment (kuisathaverat) [inline src/main/java/org/jenkinsci/plugins/saml/BundleKeyStore.java]: maybe we can save it on the Jenkins Home and make it permanent but we have to store the passwords in some place encrypted, I'm not sure how to do that.

**Comment by amuniz** [comment]
Findbugs error:
```
INFO] Redundant nullcheck of org.jenkinsci.plugins.saml.SamlSecurityRealm.getSamlPluginConfig(), which is known to be non-null in org.jenkinsci.plugins.saml.SamlSecurityRealm.toString() [org.jenkinsci.plugins.saml.SamlSecurityRealm] Redundant null check at SamlSecurityRealm.java:[line 560] RCN_REDUNDANT_NULLCHECK_OF_NONNULL_VALUE
```

Not sure why we are not getting CI build on latest commits here.

File: src/main/java/org/jenkinsci/plugins/saml/OpenSAMLWrapper.java (hunk 9 above)
Comment (amuniz) [inline src/main/java/org/jenkinsci/plugins/saml/OpenSAMLWrapper.java]: We should get a clear vision of this before merging. It is an important point IMO.

**Comment by kuisathaverat** [comment]
@amuniz it really is not a change, it was there from the first versions https://github.com/jenkinsci/saml-plugin/blob/master/src/main/java/org/jenkinsci/plugins/saml/SamlSecurityRealm.java#L394

**Comment by kuisathaverat** [comment]
in any case, I will make a test locally to see how SimpleSAML manage it.

**Comment by kuisathaverat** [comment]
SimpleSAML does not check the callback URL on the SAMLRequest and uses the configured on the SP metadata, pac4j checks it from the SAMLResponse and if the URL does not match the authentification fails, also pac4j needs a value on it and only support one value, so should be the same value that it is in the SP metadata configured on the IdP, and it makes sense because the SP metadata should come from the SP metadata provided (http://jenkins.example.com:8080/jenkins/securityRealm/metadata)

SP Metadata

```
<AssertionConsumerService> [One or More]
One or more elements that describe indexed endpoints that support the profiles of the
Authentication Request protocol defined in [SAMLProf]. All service providers support at least one
such endpoint, by definition.
```

SAMLResponse

```
AssertionConsumerServiceIndex [Optional]
Indirectly identifies the location to which the <Response> message should be returned to the
requester. It applies only to profiles in which the requester is different from the presenter, such as the
Web Browser SSO profile in [SAMLProf]. The identity provider MUST have a trusted means to map
the index value in the attribute to a location associated with the requester. [SAMLMeta] provides one
possible mechanism. If omitted, then the identity provider MUST return the <Response> message to
the default location associated with the requester for the profile of use. If the index specified is invalid,
then the identity provider MAY return an error <Response> or it MAY use the default location. This
attribute is mutually exclusive with the AssertionConsumerServiceURL and ProtocolBinding
attributes.

AssertionConsumerServiceURL [Optional]
Specifies by value the location to which the <Response> message MUST be returned to the
requester. The responder MUST ensure by some means that the value specified is in fact associated
with the requester. [SAMLMeta] provides one possible mechanism; signing the enclosing
<AuthnRequest> message is another. This attribute is mutually exclusive with the
AssertionConsumerService
```

http://docs.oasis-open.org/security/saml/v2.0/saml-core-2.0-os.pdf
http://docs.oasis-open.org/security/saml/v2.0/saml-metadata-2.0-os.pdf
